master/internal/api_experiment.go

```diff
@@ -957,67 +985,74 @@ func (a *apiServer) GetExperimentCheckpoints(
+	checkpoints := []*Checkpoint{}
+	q := db.Bun().NewSelect().
+		Model(&checkpoints).
+		Where("experiment_id = ?", req.Id)
 
-		found := false
-		for _, state := range req.States {
-			if state == v.State {
-				found = true
-				break
-			}
-		}
+	if len(req.States) > 0 {
+		states := []string{}
+		for _, s := range req.States {
+			state_text := checkpointv1.State_name[int32(s.Number())]
+			states = append(states, state_text)
+		}
+		q.Where("state in (?)", bun.In(states))
+
+	}
+
+	orderColumn := "trial_id, report_time"
+	switch req.SortBy {
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_BATCH_NUMBER:
+		orderColumn = "(metadata->>'steps_completed')::int"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_UUID:
+		orderColumn = "uuid"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_END_TIME:
+		orderColumn = "report_time"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_STATE:
+		orderColumn = "state_enum"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_SEARCHER_METRIC:
+		orderColumn = "(training->>'searcher_metric')::float8"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_TRIAL_ID:
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_UNSPECIFIED:
+	default:
+	}
 
-		if len(req.States) != 0 && !found {
-			return false
-		}
+	sortDirection := db.SortDirectionAsc
+	if req.OrderBy == apiv1.OrderBy_ORDER_BY_DESC {
+		sortDirection = db.SortDirectionDescNullsLast
+	}
 
-		return true
-	})
+	if req.Limit == 0 {
+		req.Limit = 100
+	}
 
-	sort.Slice(resp.Checkpoints, func(i, j int) bool {
-		ai, aj := resp.Checkpoints[i], resp.Checkpoints[j]
-		if useSearcherSortBy {
-			if order, done := protoless.CheckpointSearcherMetricNullsLast(ai, aj); done {
-				return order
-			}
-		}
+	db.PaginateBunUnsafe(
+		q,
+		orderColumn,
+		sortDirection,
+		int(req.Offset),
+		int(req.Limit))
 
-		if req.OrderBy == apiv1.OrderBy_ORDER_BY_DESC {
-			aj, ai = ai, aj
-		}
-
-		switch req.SortBy {
-		case apiv1.GetExperimentCheckpointsRequest_SORT_BY_BATCH_NUMBER:
-			return protoless.CheckpointStepsCompletedLess(ai, aj)
-		case apiv1.GetExperimentCheckpointsRequest_SORT_BY_UUID:
-			return ai.Uuid < aj.Uuid
-		case apiv1.GetExperimentCheckpointsRequest_SORT_BY_TRIAL_ID:
-			return protoless.CheckpointTrialIDLess(ai, aj)
-		case apiv1.GetExperimentCheckpointsRequest_SORT_BY_END_TIME:
-			return protoless.CheckpointReportTimeLess(ai, aj)
-		case apiv1.GetExperimentCheckpointsRequest_SORT_BY_STATE:
-			return ai.State.Number() < aj.State.Number()
-		case apiv1.GetExperimentCheckpointsRequest_SORT_BY_SEARCHER_METRIC:
-			return protoless.CheckpointSearcherMetricLess(ai, aj)
-		case apiv1.GetExperimentCheckpointsRequest_SORT_BY_UNSPECIFIED:
-			fallthrough
-		default:
-			return protoless.CheckpointTrialIDLess(ai, aj)
-		}
-	})
-	return resp, a.paginate(&resp.Pagination, &resp.Checkpoints, req.Offset, req.Limit)
+	err = q.Scan(context.TODO(), &checkpoints)
+
+	switch {
+	case errors.Cause(err) == sql.ErrNoRows || err == db.ErrNotFound:
```

Comment:
i usually just us `errors.Is(err, sql.ErrNoRows)`, but I guess cause is fine since it would definitely be the root error.